connectors/sources/google.py

```diff
@@ -0,0 +1,252 @@
+            )
+            raise
+        except HTTPError as exception:
+            self._logger.warning(
+                f"Response code: {exception.res.status_code} Exception: {exception}."
+            )
+            raise
+        except Exception as exception:
+            self._logger.warning(f"Exception: {exception}.")
+            raise
+
+
+def remove_universe_domain(json_credentials):
+    if "universe_domain" in json_credentials:
+        json_credentials.pop("universe_domain")
+
+
+class GoogleDirectoryClient:
+    def __init__(self, json_credentials, customer_id, timeout=DEFAULT_TIMEOUT):
+        remove_universe_domain(json_credentials)
+        self._customer_id = customer_id
+        self._client = GoogleServiceAccountClient(
+            json_credentials=json_credentials,
+            api="admin",
+            api_version="directory_v1",
+            scopes=["https://www.googleapis.com/auth/admin.directory.user.readonly"],
+            api_timeout=timeout,
+        )
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    async def ping(self):
+        try:
+            await self._client.api_call(
+                resource="users",
+                method="list",
+                maxResults=1,
+                customer=self._customer_id,
+            )
+        except Exception:
+            raise
+
+    async def users(self):
+        users_fields = f"{UserFields.EMAIL.value},{UserFields.CREATION_DATE.value}"
+
+        async for page in self._client.api_call_paged(
+            resource="users",
+            method="list",
+            fields=f"nextPageToken,users({users_fields})",
+            pageSize=DEFAULT_PAGE_SIZE,
+            customer=self._customer_id,
+        ):
+            for user in page.get("users", []):
+                yield user
+
+
+class GMailClient:
+    def __init__(self, json_credentials, customer_id, subject, timeout=DEFAULT_TIMEOUT):
+        remove_universe_domain(json_credentials)
+
+        # This override is needed to be able to fetch the messages for the corresponding user, otherwise we get a 403 Forbidden (see: https://issuetracker.google.com/issues/290567932)
+        json_credentials["subject"] = subject
+        self.user = subject
+        self._customer_id = customer_id
+        self._client = GoogleServiceAccountClient(
+            json_credentials=json_credentials,
+            api="gmail",
+            api_version="v1",
+            scopes=["https://www.googleapis.com/auth/gmail.readonly"],
+            api_timeout=timeout,
+        )
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    async def ping(self):
+        try:
+            await self._client.api_call(
+                resource="users", method="getProfile", userId=self.user
+            )
+        except Exception:
+            raise
+
+    async def messages(self, query=None, pageSize=DEFAULT_PAGE_SIZE):
+        fields = "id"
```

Comment:
Can we have all the fields needed here, so that we don't have to call method `message` for each message.

Reply:
Citing from the [GMail messages API](https://developers.google.com/gmail/api/reference/rest/v1/users.messages/list):

> List of messages. Note that each message resource contains only an id and a threadId. Additional message details can be fetched using the [messages.get](https://developers.google.com/gmail/api/v1/reference/users/messages/get) method.

Reply:
Noted. Probably we can consider batching the get calls: https://developers.google.com/gmail/api/guides/batch